feat: add support for replacing the substring before the first occurrence of a search string (#843)
stdlib-js/stdlib: lib/node_modules/@stdlib/string/base/replace-before/README.md

```

</section>

<!-- /.usage -->

<!-- CLI usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

### Notes

-   If the split separator is a [regular expression][mdn-regexp], ensure that the `split` option is either properly escaped or enclosed in quotes.

    ```bash
    # Not escaped...
    $ echo -n $'foo\nbar\nbaz' | replace-before --search a --replacement b --split /\r?\n/

    # Escaped...
    $ echo -n $'foo\nbar\nbaz' | replace-before --search a --replacement b --split /\\r?\\n/
    ```

-   The implementation ignores trailing delimiters.

</section>

<!-- /.notes -->

<!-- CLI usage examples. -->

<section class="examples">

### Examples

```bash
$ replace-before abcdefg --search d --replacement pqr
pqrdefg
```

To use as a [standard stream][standard-streams],

```bash
$ echo -n $'beep\nboop' | replace-before --search p --replacement see
seep 
seep
```

By default, when used as a [standard stream][standard-streams], the implementation assumes newline-delimited data. To specify an alternative delimiter, set the `split` option.

```bash

> $ echo -n 'beep\tboop' | replace-before --search p --see --split '\t'

Reviewer: ```suggestion
$ echo -n 'beep\tboop' | replace-before --search p --replacement see --split '\t'
```

<!--

@license Apache-2.0

> Copyright (c) 2021 The Stdlib Authors.

Reviewer 2: ```suggestion
Copyright (c) 2023 The Stdlib Authors.
```

You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# replaceBefore

> Replace the substring before the first occurrence of a specified search string.

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- Package usage documentation. -->

<section class="usage">

## Usage

```javascript
var replaceBefore = require( '@stdlib/string/base/replace-before' );
```

#### replaceBefore( str, search, replacement )

Replaces the substring before the first occurrence of a specified search string.

```javascript
var str = 'beep boop';
var replacement = 'loop';
var out = replaceBefore( str, ' ', replacement );
// returns 'loop boop'

replacement = 'bar';

> out = replaceBefore( str, 'o' , replacement);

Reviewer 2: ```suggestion
out = replaceBefore( str, 'o' , replacement );
```

> var str = 'beep boop';
> var replacement = 'loop';
> var out = replaceBefore( str, ' ', replacement );
> // returns 'loop boop'
> 
> replacement = 'bar';
> out = replaceBefore( str, 'o' , replacement );
> // returns 'baroop'

Reviewer: ```suggestion
var out = replaceBefore( 'beep boop', ' ', 'loop' );
// returns 'loop boop'

out = replaceBefore( 'beep boop', 'o', 'bar' );
// returns 'baroop'
```

Reviewer: We can inline the variable values to reduce "noise" in the example without loss of clarity.

```

</section>

<!-- /.usage -->

<!-- Package usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

## Notes

> -   If a substring is not present in a provided string, the function returns the input string.

Reviewer: ```suggestion
-   If a search string is not present in a provided string, the function returns the provided string unchanged.
```

Reviewer: We can use parallel sentence construction to avoid ambiguity.

> -   If provided an empty substring, the function returns an empty string.

Reviewer: ```suggestion
```

Reviewer: Not clear what info this is intended to convey.

</section>

<!-- /.notes -->

<!-- Package usage examples. -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript

> var substringBefore = require( '@stdlib/string/base/replace-before' );

Reviewer: s/substringBefore/replaceBefore/

This looks to be a copy-paste error.

var out = substringBefore( 'beep boop', 'p', 'see' );
// returns 'seep boop'

out = substringBefore( 'Hello World!', 'xyz', 'foo' );

> // returns 'fooHello World!'

Reviewer: This looks wrong to me. If the search string is not present, we should return the input string unchanged.

> out = substringBefore( 'Hello World!', '', 'foo' );

Reviewer: IMO, if the search string is an empty string, we should return the provided string unchanged. Prepending, as in this example, does not seem intuitive.

// returns 'fooHello World!'

> out = substringBefore( '', 'xyz' , 'foo');

Reviewer: This is also unexpected behavior. If the search string is not present, I would not expect the replacement string to be present in the return string.

// returns 'foo'
```

</section>

<!-- /.examples -->

<!-- Section for describing a command-line interface. -->

* * *

<section class="cli">

## CLI

<!-- CLI usage documentation. -->

<section class="usage">

### Usage

```text

> Usage: replace-before [options] --search=<string> [<string>] --replacement=<string> [<string>]

Reviewer: ```suggestion
Usage: replace-before [options] --search=<string> --replacement=<string> [<string>]
```

Options:

  -h,    --help                Print this message.
  -V,    --version             Print the package version.
         --search string       Search string.
         --split sep           Delimiter for stdin data. Default: '/\\r?\\n/'.
```

</section>

<!-- /.usage -->

<!-- CLI usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

### Notes

-   If the split separator is a [regular expression][mdn-regexp], ensure that the `split` option is either properly escaped or enclosed in quotes.

    ```bash
    # Not escaped...
    $ echo -n $'foo\nbar\nbaz' | replace-before --search a --split /\r?\n/

    # Escaped...

>     $ echo -n $'foo\nbar\nbaz' | replace-before --search a --split /\\r?\\n/

Reviewer: Based on L118, the `replacement` option is required. Yet, this example does not include it in either L142 or L145.

    ```

-   The implementation ignores trailing delimiters.

</section>

<!-- /.notes -->

<!-- CLI usage examples. -->

<section class="examples">

### Examples

```bash
$ replace-before abcdefg --search d --replacement pqr
pqrdefg
```

To use as a [standard stream][standard-streams],

```bash
$ echo -n $'beep\nboop' | replace-before --search p --replacement see
seep boop

> beep seep

Reviewer: The return values are incorrect. Should be `seep` for both lines.

```

By default, when used as a [standard stream][standard-streams], the implementation assumes newline-delimited data. To specify an alternative delimiter, set the `split` option.

```bash
$ echo -n 'beep\tboop' | replace-before --search p --see --split '\t'
seep boop

> beep seep

Reviewer: Same comment as above.

```

</section>

<!-- /.examples -->

</section>

<!-- /.cli -->

<!-- Section to include cited references. If references are included, add a horizontal rule *before* the section. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="references">

</section>

<!-- /.references -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

> * * *
> 
> ## See Also
> 
> -   <span class="package-name">[`@stdlib/string/substring-before-last`][@stdlib/string/substring-before-last]</span><span class="delimiter">: </span><span class="description">return the part of a string before the last occurrence of a specified substring.</span>
> -   <span class="package-name">[`@stdlib/string/substring-after`][@stdlib/string/substring-after]</span><span class="delimiter">: </span><span class="description">return the part of a string after a specified substring.</span>
> -   <span class="package-name">[`@stdlib/string/substring-after-last`][@stdlib/string/substring-after-last]</span><span class="delimiter">: </span><span class="description">return the part of a string after the last occurrence of a specified substring.</span>
> 

Reviewer: ```suggestion
```

The list of related packages is done by an automatic build process and should not be included here.

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[standard-streams]: https://en.wikipedia.org/wiki/Standard_streams

[mdn-regexp]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions

> <!-- <related-links> -->
> 
> [@stdlib/string/substring-before-last]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/string/substring-before-last
> 
> [@stdlib/string/substring-after]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/string/substring-after
> 
> [@stdlib/string/substring-after-last]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/string/substring-after-last
> 
> <!-- </related-links> -->
> 

Reviewer: ```suggestion
```

Can remove as handled by build process.
